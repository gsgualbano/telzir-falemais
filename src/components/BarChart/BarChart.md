BarChart Exemplo:

```js
const props = {
  formData: {
    origem: "11",
    destino: "16",
    tempo: 20,
    precos: { comFaleMais: 0.0, semFaleMais: 38.0 },
    plano: "FaleMais 30",
    economia: "100%"
  },
  isVisible: true,
  setShowChart: () => {}
};

<BarChart {...props} />;
```
