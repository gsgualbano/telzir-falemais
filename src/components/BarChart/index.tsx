import React from 'react';

import Chart from 'react-apexcharts';
import { faLongArrowAltLeft } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

import { Animated } from 'react-animated-css';

import {
  Container, Data, Label, Description, DataGroup, Text, Footer, BackButton, BackButtonText,
} from './styles';


interface OwnProps {
  /** Dados que vem do formulário no Body como origem, destino, tempo, plano, preços, economia */
  formData: {
    origem: string,
    destino: string,
    tempo: number,
    plano: string,
    precos: { comFaleMais: number, semFaleMais: number },
    economia: string,
  },
  /** Valor booleano para que indica se o componente está visível ou não */
  isVisible: boolean,
  /** Função dispatch que atualizar o valor booleano do IsVisible */
  setShowChart: React.Dispatch<React.SetStateAction<boolean>>
}

/**
 * Gráfico em Barra e dados sobre as tarifas
 */
const BarChart = ({ formData, isVisible, setShowChart }: OwnProps) => {
  const {
    origem,
    destino,
    tempo,
    plano,
    precos,
    economia,
  } = formData;

  const colors = ['#00E396', '#FF0000'];

  /** Configurações do gráfico */
  const options = {
    chart: {
      height: 350,
      type: 'bar',
      events: {},
    },
    colors,
    plotOptions: {
      bar: {
        columnWidth: '45%',
        distributed: true,
      },
    },
    dataLabels: {
      enabled: false,
    },
    xaxis: {
      categories: ['Com FaleMais', 'Sem FaleMais'],
      labels: {
        style: {
          colors,
          fontSize: '14px',
        },
      },
    },
  };

  /** Valores do gráfico */
  const series = [
    {
      name: 'Preço R$',
      data: [Number(precos.comFaleMais.toFixed(2)), Number(precos.semFaleMais.toFixed(2))],
    },
  ];

  return (
    <Animated animationIn="fadeIn" animationOut="fadeOut" isVisible={isVisible}>
      <Container>
        <BackButton onClick={() => setShowChart(false)}>
          <FontAwesomeIcon icon={faLongArrowAltLeft} />
          <BackButtonText>voltar</BackButtonText>
        </BackButton>
        <Data>
          <DataGroup>
            <Label>DDD Origem:</Label>
            <Description data-testid="origem">{origem}</Description>
          </DataGroup>
          <DataGroup>
            <Label>DDD Destino:</Label>
            <Description data-testid="destino">{destino}</Description>
          </DataGroup>
          <DataGroup>
            <Label>Tempo:</Label>
            <Description data-testid="tempo">{`${tempo} minutos`}</Description>
          </DataGroup>
          <DataGroup>
            <Label>Plano:</Label>
            <Description data-testid="plano">{plano}</Description>
          </DataGroup>
          <DataGroup fontSize="25px">
            <Label color="#00E396" fontFamily="Lobster">FaleMais:</Label>
            <Description data-testid="comFaleMais" color="#00E396">
              {`R$ ${!Number.isNaN(precos.comFaleMais) ? precos.comFaleMais.toFixed(2) : '-'}`}
            </Description>
          </DataGroup>
          <DataGroup>
            <Label color="#ff0000">Sem FaleMais:</Label>
            <Description data-testid="semFaleMais" color="#ff0000">
              {`R$ ${!Number.isNaN(precos.semFaleMais) ? precos.semFaleMais.toFixed(2) : '-'}`}
            </Description>
          </DataGroup>
        </Data>
        <Chart options={options} series={series} type="bar" width="100%" />
        <Footer>
          <Text>
            Com o FaleMais você economiza:
            <span style={{ color: '#00E396' }} data-testid="economia">{economia}</span>
          </Text>
        </Footer>
      </Container>
    </Animated>
  );
};

export default BarChart;
