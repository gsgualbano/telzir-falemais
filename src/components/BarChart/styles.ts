import styled from 'styled-components';

interface LabelProps {
  color?: string;
  fontFamily?: string;
  fontSize?: string;
}

export const Container = styled.div``;

export const BackButton = styled.button`
  display: flex;
  color: #3C40C6;
  font-size: 18px;
  padding: 5px;
  border: none;
  outline: none;
  justify-content: space-between;
  background: #fff;
  align-items: center;

  &:hover {
    color: #242AC6;
  }
`;

export const BackButtonText = styled.span`
  margin-left: 2px;
`;

export const Data = styled.div`
  display: flex;
  flex-direction: column;
`;

export const DataGroup = styled.div`
  display: flex;
  justify-content: space-between;
  padding: 5px;
  font-size: ${({ fontSize }: LabelProps) => (fontSize || '16px')};

  &:last-child { //ultimo
    padding-bottom: 10px;
    border-bottom: 1px solid #ebebeb;
  }
`;

export const Label = styled.div`
  color: ${({ color }: LabelProps) => (color || '#8b8b8b')};
  font-weight: 600;
  font-family: ${({ fontFamily }: LabelProps) => (fontFamily || "'Open Sans', sans-serif")}
`;

export const Description = styled.div`
  color: ${({ color }: LabelProps) => (color || '#000000')};
`;

export const Footer = styled.div`
  border-top: 1px solid #ebebeb;
  padding: 5px;
`;

export const Text = styled.p`
  text-align: center;
`;
