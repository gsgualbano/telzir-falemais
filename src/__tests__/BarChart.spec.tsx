import React from 'react';
import BarChart from '../components/BarChart';

import 'jest-dom/extend-expect';
import { shallow } from '../enzyme';

describe('<BarChart />', () => {
  const props = {
    formData: {
      origem: '11',
      destino: '18',
      tempo: 80,
      precos: { comFaleMais: 40.28, semFaleMais: 60.40 },
      plano: 'FaleMais 30',
      economia: '80%',
    },
    isVisible: true,
    setShowChart: () => { },
  };

  it('renders correctly', () => {
    const wrapper = shallow(<BarChart {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('shows correct data', () => {
    const wrapper = shallow(<BarChart {...props} />);
    expect(wrapper.find("[data-testid='origem']").text()).toEqual('11');
    expect(wrapper.find("[data-testid='destino']").text()).toEqual('18');
    expect(wrapper.find("[data-testid='tempo']").text()).toEqual('80 minutos');
    expect(wrapper.find("[data-testid='comFaleMais']").text()).toEqual('R$ 40.28');
    expect(wrapper.find("[data-testid='semFaleMais']").text()).toEqual('R$ 60.40');
    expect(wrapper.find("[data-testid='plano']").text()).toEqual('FaleMais 30');
    expect(wrapper.find("[data-testid='economia']").text()).toEqual('80%');
  });

  it('shows correct data when passing undefined props', () => {
    const propsUndefined = {
      formData: {
        origem: '18',
        destino: '17',
        tempo: 10,
        precos: { comFaleMais: NaN, semFaleMais: NaN },
        plano: 'FaleMais 120',
        economia: '-',
      },
      isVisible: true,
      setShowChart: () => { },
    };

    const wrapper = shallow(<BarChart {...propsUndefined} />);
    expect(wrapper.find("[data-testid='origem']").text()).toEqual('18');
    expect(wrapper.find("[data-testid='destino']").text()).toEqual('17');
    expect(wrapper.find("[data-testid='tempo']").text()).toEqual('10 minutos');
    expect(wrapper.find("[data-testid='comFaleMais']").text()).toEqual('R$ -');
    expect(wrapper.find("[data-testid='semFaleMais']").text()).toEqual('R$ -');
    expect(wrapper.find("[data-testid='plano']").text()).toEqual('FaleMais 120');
    expect(wrapper.find("[data-testid='economia']").text()).toEqual('-');
  });
});
