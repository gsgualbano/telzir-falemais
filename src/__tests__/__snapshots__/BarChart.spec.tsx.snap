// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BarChart /> renders correctly 1`] = `
<Animated
  animateOnMount={true}
  animationIn="fadeIn"
  animationInDelay={0}
  animationInDuration={1000}
  animationOut="fadeOut"
  animationOutDelay={0}
  animationOutDuration={1000}
  className=""
  isVisible={true}
  style={Object {}}
>
  <styled.div>
    <styled.button
      onClick={[Function]}
    >
      <FontAwesomeIcon
        border={false}
        className=""
        fixedWidth={false}
        flip={null}
        icon={
          Object {
            "icon": Array [
              448,
              512,
              Array [],
              "f30a",
              "M134.059 296H436c6.627 0 12-5.373 12-12v-56c0-6.627-5.373-12-12-12H134.059v-46.059c0-21.382-25.851-32.09-40.971-16.971L7.029 239.029c-9.373 9.373-9.373 24.569 0 33.941l86.059 86.059c15.119 15.119 40.971 4.411 40.971-16.971V296z",
            ],
            "iconName": "long-arrow-alt-left",
            "prefix": "fas",
          }
        }
        inverse={false}
        listItem={false}
        mask={null}
        pull={null}
        pulse={false}
        rotation={null}
        size={null}
        spin={false}
        symbol={false}
        title=""
        transform={null}
      />
      <styled.span>
        voltar
      </styled.span>
    </styled.button>
    <styled.div>
      <styled.div>
        <styled.div>
          DDD Origem:
        </styled.div>
        <styled.div
          data-testid="origem"
        >
          11
        </styled.div>
      </styled.div>
      <styled.div>
        <styled.div>
          DDD Destino:
        </styled.div>
        <styled.div
          data-testid="destino"
        >
          18
        </styled.div>
      </styled.div>
      <styled.div>
        <styled.div>
          Tempo:
        </styled.div>
        <styled.div
          data-testid="tempo"
        >
          80 minutos
        </styled.div>
      </styled.div>
      <styled.div>
        <styled.div>
          Plano:
        </styled.div>
        <styled.div
          data-testid="plano"
        >
          FaleMais 30
        </styled.div>
      </styled.div>
      <styled.div
        fontSize="25px"
      >
        <styled.div
          color="#00E396"
          fontFamily="Lobster"
        >
          FaleMais:
        </styled.div>
        <styled.div
          color="#00E396"
          data-testid="comFaleMais"
        >
          R$ 40.28
        </styled.div>
      </styled.div>
      <styled.div>
        <styled.div
          color="#ff0000"
        >
          Sem FaleMais:
        </styled.div>
        <styled.div
          color="#ff0000"
          data-testid="semFaleMais"
        >
          R$ 60.40
        </styled.div>
      </styled.div>
    </styled.div>
    <r
      height="auto"
      options={
        Object {
          "chart": Object {
            "events": Object {},
            "height": 350,
            "type": "bar",
          },
          "colors": Array [
            "#00E396",
            "#FF0000",
          ],
          "dataLabels": Object {
            "enabled": false,
          },
          "plotOptions": Object {
            "bar": Object {
              "columnWidth": "45%",
              "distributed": true,
            },
          },
          "xaxis": Object {
            "categories": Array [
              "Com FaleMais",
              "Sem FaleMais",
            ],
            "labels": Object {
              "style": Object {
                "colors": Array [
                  "#00E396",
                  "#FF0000",
                ],
                "fontSize": "14px",
              },
            },
          },
        }
      }
      series={
        Array [
          Object {
            "data": Array [
              40.28,
              60.4,
            ],
            "name": "Preço R$",
          },
        ]
      }
      type="bar"
      width="100%"
    />
    <styled.div>
      <styled.p>
        Com o FaleMais você economiza:
        <span
          data-testid="economia"
          style={
            Object {
              "color": "#00E396",
            }
          }
        >
          80%
        </span>
      </styled.p>
    </styled.div>
  </styled.div>
</Animated>
`;
